File: realtorcom-scraper/browser/python/realtorcom.py
"""Realtor.com scraper using the official Scrapeless Python SDK + Playwright over CDP.
function names + emitted field names match verbatim.

Three surfaces:
- `scrape_property(url)`       — extracts `__NEXT_DATA__` JSON, reduces via JMESPath.
- `scrape_search(state, city, max_pages)` — `__NEXT_DATA__` array of property cards.
- `scrape_feed(url)`           — XML sitemap with `<loc>` + `<lastmod>` pairs.
"""

from __future__ import annotations

import asyncio
import json
import math
import os
from datetime import datetime
from typing import Any

import jmespath
from loguru import logger
from parsel import Selector
from playwright.async_api import async_playwright
from scrapeless import Scrapeless
from scrapeless.types import ICreateBrowser

DEFAULT_PROXY_COUNTRY = "US"
DEFAULT_SESSION_TTL = 180

def _client() -> Scrapeless:
    if not os.environ.get("SCRAPELESS_API_KEY") and not os.environ.get("SCRAPELESS_KEY"):
        raise RuntimeError(
            "SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com"
        )
    if not os.environ.get("SCRAPELESS_API_KEY") and os.environ.get("SCRAPELESS_KEY"):
        os.environ["SCRAPELESS_API_KEY"] = os.environ["SCRAPELESS_KEY"]
    return Scrapeless()

async def _new_browser(proxy_country: str = DEFAULT_PROXY_COUNTRY) -> str:
    client = _client()
    session = client.browser.create(
        ICreateBrowser(proxy_country=proxy_country, session_ttl=DEFAULT_SESSION_TTL)
    )
    return session.browser_ws_endpoint

async def _fetch_html(
    url: str,
    ready_selector: str | None = None,
    *,
    retries: int = 2,
    warmup: bool = True,
) -> str:
    """Realtor.com is fronted by PerimeterX; a homepage warm-up gets the
    `_px*` cookies before navigating to the search / property page.
    """
    last_error: Exception | None = None
    for attempt in range(retries + 1):
        if attempt > 0:
            await asyncio.sleep(6 * attempt)
        ws = await _new_browser()
        async with async_playwright() as pw:
            browser = await pw.chromium.connect_over_cdp(ws)
            try:
                page = await browser.new_page()
                if warmup and not url.endswith(".xml"):
                    try:
                        await page.goto("https://www.realtor.com/", wait_until="domcontentloaded", timeout=30000)
                        await asyncio.sleep(3)
                    except Exception:
                        pass
                await page.goto(url, wait_until="domcontentloaded", timeout=45000)
                if ready_selector:
                    try:
                        await page.wait_for_selector(ready_selector, timeout=15000)
                    except Exception as e:
                        logger.warning("wait_for_selector failed: {}", e)
                html = await page.content()
                if html and (url.endswith(".xml") or len(html) > 10000):
                    return html
                last_error = RuntimeError("PerimeterX interstitial / short HTML")
            except Exception as e:  # noqa: BLE001
                last_error = e
                logger.warning("attempt {} failed: {}", attempt + 1, e)
            finally:
                try:
                    await browser.close()
                except Exception:
                    pass
    raise RuntimeError(f"giving up after {retries + 1} attempts: {last_error}")

# ---------------------------------------------------------------------------
# Parsers — mirror the upstream reference verbatim
# ---------------------------------------------------------------------------

PROPERTY_JMESPATH = """{
    id: propertyDetails.listing_id,
    slug: slug,
    url: propertyDetails.href,
    status: propertyDetails.status,
    tags: propertyDetails.tags,
    sold_date: propertyDetails.last_sold_date,
    sold_price: propertyDetails.last_sold_price,
    list_date: propertyDetails.list_date,
    list_price: propertyDetails.list_price,
    list_price_last_change: propertyDetails.last_price_change_amount,
    details: propertyDetails.description,
    flags: propertyDetails.flags,
    local: propertyDetails.local,
    location: propertyDetails.location,
    agent: propertyDetails.source.agents,
    advertisers: propertyDetails.advertisers,
    tax_history: propertyDetails.tax_history,
    history: propertyDetails.property_history[].{
        date: date,
        event: event_name,
        price: price,
        price_sqft: price_sqft
    },
    photos: propertyDetails.photos[].{
        url: href,
        tags: tags[].label
    },
    phones: propertyDetails.lead_attributes.opcity_lead_attributes.phones[].{
        type: category,
        number: number
    },
    features: propertyDetails.details[].{
        name: category,
        values: text
    }
}"""

def parse_property(html: str, url: str) -> dict | None:
    """Returns the JMESPath-reduced property dict."""
    sel = Selector(text=html)
    data = sel.css("script#__NEXT_DATA__::text").get()
    if not data:
        logger.warning("page {} is not a property listing page", url)
        return None
    parsed = json.loads(data)
    raw_data = parsed["props"]["pageProps"]["initialReduxState"]
    reduced = jmespath.search(PROPERTY_JMESPATH, raw_data)
    if reduced and reduced.get("features"):
        reduced["features"] = {f["name"]: f["values"] for f in reduced["features"] if f.get("name")}
    return reduced

def parse_search(html: str, url: str) -> dict | None:
    """"""
    sel = Selector(text=html)
    data = sel.css("script#__NEXT_DATA__::text").get()
    if not data:
        logger.warning("page {} is not a property listing page", url)
        return None
    parsed = json.loads(data)["props"]["pageProps"]
    if not parsed.get("properties"):
        parsed["properties"] = parsed["searchResults"]["home_search"]["results"]
    if not parsed.get("totalProperties"):
        parsed["totalProperties"] = parsed["searchResults"]["home_search"]["total"]
    return parsed

# ---------------------------------------------------------------------------
# Scrape functions — mirror the upstream reference verbatim
# ---------------------------------------------------------------------------

async def scrape_property(url: str) -> dict | None:
    """Scrape realtor.com's property page for property data."""
    logger.info("scraping {} property page", url)
    html = await _fetch_html(url, ready_selector="script#__NEXT_DATA__")
    return parse_property(html, url)

async def scrape_search(state: str, city: str, max_pages: int | None = None) -> list[dict]:
    """Scrape realtor.com's search and find properties for given query."""
    logger.info("scraping first property search page for {}, {}", city, state)
    first_url = f"https://www.realtor.com/realestateandhomes-search/{city}_{state}/pg-1"
    first_html = await _fetch_html(first_url, ready_selector="script#__NEXT_DATA__")
    first_data = parse_search(first_html, first_url)
    if not first_data:
        return []
    results: list[dict] = list(first_data["properties"])
    total_pages = math.ceil(first_data["totalProperties"] / len(results)) if results else 1
    if max_pages and total_pages > max_pages:
        total_pages = max_pages
    logger.info("found {} total pages", total_pages)
    for page in range(2, total_pages + 1):
        page_url = first_url.replace("pg-1", f"pg-{page}")
        try:
            html = await _fetch_html(page_url, ready_selector="script#__NEXT_DATA__")
            parsed = parse_search(html, page_url)
            if parsed:
                results.extend(parsed["properties"])
        except Exception as e:
            logger.warning("failed page {}: {}", page, e)
    logger.info("scraped search of {} results for {}, {}", len(results), city, state)
    return results

async def scrape_feed(url: str) -> dict[str, datetime]:
    """Scrape Realtor.com's sitemap-style atom feed -> { url: datetime }."""
    body = await _fetch_html(url)
    sel = Selector(text=body)
    out: dict[str, datetime] = {}
    for item in sel.xpath("//sitemap"):
        loc = item.xpath("loc/text()").get()
        pub = item.xpath("lastmod/text()").get()
        if loc and pub:
            try:
                out[loc] = datetime.fromisoformat(pub)
            except ValueError:
                pass
    return out

def to_dict(obj: Any) -> Any:
    if isinstance(obj, list):
        return [to_dict(x) for x in obj]
    if isinstance(obj, dict):
        return {k: to_dict(v) for k, v in obj.items()}
    if isinstance(obj, datetime):
        return obj.isoformat()
    return obj

File: realtorcom-scraper/browser/nodejs/realtorcom.mjs
// Realtor.com scraper using the official @scrapeless-ai/sdk + puppeteer-core over CDP.
//
// function names + emitted field names match verbatim.
//
// Three surfaces:
// - `scrapeProperty(url)`                     — `__NEXT_DATA__` JSON, JMESPath-reduced.
// - `scrapeSearch(state, city, maxPages)`     — `__NEXT_DATA__` array of property cards.
// - `scrapeFeed(url)`                         — XML sitemap with `<loc>` + `<lastmod>` pairs.

import { Scrapeless } from "@scrapeless-ai/sdk";
import puppeteer from "puppeteer-core";
import * as cheerio from "cheerio";
import jmespath from "jmespath";

const DEFAULT_PROXY_COUNTRY = "US";
const DEFAULT_SESSION_TTL = 180;

function requireKey() {
  const k = process.env.SCRAPELESS_API_KEY ?? process.env.SCRAPELESS_KEY;
  if (!k) throw new Error("SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com");
  return k;
}

function client() {
  return new Scrapeless({ apiKey: requireKey() });
}

async function newBrowser(proxyCountry = DEFAULT_PROXY_COUNTRY) {
  const { browserWSEndpoint } = await client().browser.create({
    proxyCountry,
    sessionTTL: DEFAULT_SESSION_TTL,
  });
  return browserWSEndpoint;
}

async function fetchHtml(url, readySelector = null, { retries = 2, warmup = true } = {}) {
  let lastError;
  for (let attempt = 0; attempt <= retries; attempt++) {
    if (attempt > 0) await new Promise((r) => setTimeout(r, 6000 * attempt));
    const ws = await newBrowser();
    let browser;
    try {
      browser = await puppeteer.connect({ browserWSEndpoint: ws });
      const page = await browser.newPage();
      // Realtor.com is fronted by PerimeterX. A homepage warm-up lets the
      // session pick up the `_px*` cookies before navigating to the search
      // / property page, which avoids the static "Press & Hold" interstitial.
      if (warmup && !url.endsWith(".xml")) {
        try {
          await page.goto("https://www.realtor.com/", { waitUntil: "domcontentloaded", timeout: 30000 });
          await new Promise((r) => setTimeout(r, 3000));
        } catch (_) {}
      }
      await page.goto(url, { waitUntil: "domcontentloaded", timeout: 45000 });
      if (readySelector) {
        try { await page.waitForSelector(readySelector, { timeout: 15000 }); } catch (_) {}
      }
      const html = await page.content();
      // PerimeterX interstitials are ~2KB; real pages are far larger.
      if (html && (url.endsWith(".xml") || html.length > 10000)) return html;
      lastError = new Error("PerimeterX interstitial / short HTML");
    } catch (e) {
      lastError = e;
    } finally {
      try { await browser?.close(); } catch (_) {}
    }
  }
  throw new Error(`giving up after ${retries + 1} attempts: ${lastError?.message}`);
}

const PROPERTY_JMESPATH = `{
    id: propertyDetails.listing_id,
    slug: slug,
    url: propertyDetails.href,
    status: propertyDetails.status,
    tags: propertyDetails.tags,
    sold_date: propertyDetails.last_sold_date,
    sold_price: propertyDetails.last_sold_price,
    list_date: propertyDetails.list_date,
    list_price: propertyDetails.list_price,
    list_price_last_change: propertyDetails.last_price_change_amount,
    details: propertyDetails.description,
    flags: propertyDetails.flags,
    local: propertyDetails.local,
    location: propertyDetails.location,
    agent: propertyDetails.source.agents,
    advertisers: propertyDetails.advertisers,
    tax_history: propertyDetails.tax_history,
    history: propertyDetails.property_history[].{
        date: date,
        event: event_name,
        price: price,
        price_sqft: price_sqft
    },
    photos: propertyDetails.photos[].{
        url: href,
        tags: tags[].label
    },
    phones: propertyDetails.lead_attributes.opcity_lead_attributes.phones[].{
        type: category,
        number: number
    },
    features: propertyDetails.details[].{
        name: category,
        values: text
    }
}`;

// ---------------- parsers (mirror the upstream reference verbatim) ----------------

export function parseProperty(html, url) {
  const $ = cheerio.load(html);
  const data = $("script#__NEXT_DATA__").first().text();
  if (!data) {
    console.warn(`page ${url} is not a property listing page`);
    return null;
  }
  const parsed = JSON.parse(data);
  const raw = parsed.props.pageProps.initialReduxState;
  const reduced = jmespath.search(raw, PROPERTY_JMESPATH);
  if (reduced && Array.isArray(reduced.features)) {
    const flat = {};
    for (const f of reduced.features) {
      if (f && f.name) flat[f.name] = f.values;
    }
    reduced.features = flat;
  }
  return reduced;
}

export function parseSearch(html, url) {
  const $ = cheerio.load(html);
  const data = $("script#__NEXT_DATA__").first().text();
  if (!data) {
    console.warn(`page ${url} is not a property listing page`);
    return null;
  }
  const parsed = JSON.parse(data).props.pageProps;
  if (!parsed.properties) {
    parsed.properties = parsed.searchResults.home_search.results;
  }
  if (!parsed.totalProperties) {
    parsed.totalProperties = parsed.searchResults.home_search.total;
  }
  return parsed;
}

// ---------------- scrape functions (mirror the upstream reference verbatim) ----------------

export async function scrapeProperty(url) {
  const html = await fetchHtml(url, "script#__NEXT_DATA__");
  return parseProperty(html, url);
}

export async function scrapeSearch(state, city, maxPages = null) {
  const firstUrl = `https://www.realtor.com/realestateandhomes-search/${city}_${state}/pg-1`;
  const firstHtml = await fetchHtml(firstUrl, "script#__NEXT_DATA__");
  const first = parseSearch(firstHtml, firstUrl);
  if (!first) return [];
  const results = [...first.properties];
  let totalPages = results.length > 0 ? Math.ceil(first.totalProperties / results.length) : 1;
  if (maxPages && totalPages > maxPages) totalPages = maxPages;
  for (let p = 2; p <= totalPages; p++) {
    const pageUrl = firstUrl.replace("pg-1", `pg-${p}`);
    try {
      const html = await fetchHtml(pageUrl, "script#__NEXT_DATA__");
      const parsed = parseSearch(html, pageUrl);
      if (parsed) results.push(...parsed.properties);
    } catch (e) {
      console.warn(`failed page ${p}: ${e.message}`);
    }
  }
  return results;
}

export async function scrapeFeed(url) {
  const body = await fetchHtml(url);
  const $ = cheerio.load(body, { xmlMode: true });
  const out = {};
  $("sitemap").each((_, el) => {
    const loc = $(el).find("loc").text();
    const pub = $(el).find("lastmod").text();
    if (loc && pub) {
      const d = new Date(pub);
      if (!isNaN(d.getTime())) out[loc] = d.toISOString();
    }
  });
  return out;
}
